File: apps/ios/Blackbox/Sources/BlackboxEngine.swift
import Foundation

public enum BlackboxEngineError: Error, LocalizedError {
    case allocFailed
    case engineCreateFailed(String)
    case callFailed(String)
    case outputTooLarge

    public var errorDescription: String? {
        switch self {
        case .allocFailed:
            return "blackbox bb_alloc failed"
        case .engineCreateFailed(let message):
            return message
        case .callFailed(let message):
            return message
        case .outputTooLarge:
            return "blackbox output exceeded maximum buffer size"
        }
    }
}

public final class BlackboxEngine {
    private let handle: UInt32

    public static func initialize() {
        bb_init()
    }

    public init(
        scenario: Data,
        items: Data,
        characters: Data,
        assets: Data,
        chapters: [Data],
        library: Data? = nil,
        randomSeedOverride: UInt64? = nil
    ) throws {
        let scenarioPtr = try Self.copyIntoRust(scenario)
        let itemsPtr = try Self.copyIntoRust(items)
        let charactersPtr = try Self.copyIntoRust(characters)
        let assetsPtr = try Self.copyIntoRust(assets)
        let libraryPtr = try Self.copyIntoRust(library ?? Data())

        var chapterSlices = [ByteSlice]()
        var chapterPtrs: [(ptr: Int, len: UInt32)] = []
        for chapter in chapters {
            let copied = try Self.copyIntoRust(chapter)
            chapterPtrs.append((copied.ptr, copied.len))
            chapterSlices.append(ByteSlice(ptr: copied.ptr, len: copied.len))
        }

        let tablePtr: Int
        if chapterSlices.isEmpty {
            tablePtr = 0
        } else {
            tablePtr = try chapterSlices.withUnsafeBytes { raw in
                let len = UInt32(raw.count)
                let ptr = bb_alloc(len)
                guard ptr != 0 else {
                    throw BlackboxEngineError.allocFailed
                }
                memcpy(UnsafeMutableRawPointer(bitPattern: ptr)!, raw.baseAddress!, raw.count)
                return ptr
            }
        }

        defer {
            Self.freeCopy(scenarioPtr)
            Self.freeCopy(itemsPtr)
            Self.freeCopy(charactersPtr)
            Self.freeCopy(assetsPtr)
            Self.freeCopy(libraryPtr)
            for copied in chapterPtrs {
                bb_free(copied.ptr, copied.len)
            }
            if tablePtr != 0 {
                bb_free(tablePtr, UInt32(MemoryLayout<ByteSlice>.stride * chapterSlices.count))
            }
        }

        let created = bb_engine_new(
            scenarioPtr.ptr,
            scenarioPtr.len,
            itemsPtr.ptr,
            itemsPtr.len,
            charactersPtr.ptr,
            charactersPtr.len,
            assetsPtr.ptr,
            assetsPtr.len,
            tablePtr,
            UInt32(chapters.count),
            libraryPtr.ptr,
            libraryPtr.len,
            randomSeedOverride != nil,
            randomSeedOverride ?? 0
        )

        guard created != 0 else {
            throw BlackboxEngineError.engineCreateFailed(Self.readLastError())
        }
        handle = created
    }

    deinit {
        bb_engine_free(handle)
    }

    public func getCurrentView() throws -> String {
        try String(decoding: callWithOutput { outPtr, outCap in
            bb_get_view(handle, outPtr, outCap)
        }, as: UTF8.self)
    }

    public func submitCommand(_ json: String, viewRevision: UInt32) throws -> String {
        try String(decoding: callWithUtf8Input(json) { inPtr, inLen, outPtr, outCap in
            bb_submit(handle, inPtr, inLen, viewRevision, outPtr, outCap)
        }, as: UTF8.self)
    }

    public func loadCatalog(_ data: Data) throws {
        try callWithDataInput(data) { inPtr, inLen in
            bb_load_catalog(handle, inPtr, inLen)
        }
    }

    public func loadLibrary(_ data: Data) throws {
        try callWithDataInput(data) { inPtr, inLen in
            bb_load_library(handle, inPtr, inLen)
        }
    }

    public func loadChapter(_ data: Data) throws {
        try callWithDataInput(data) { inPtr, inLen in
            bb_load_chapter(handle, inPtr, inLen)
        }
    }

    public func unloadChapter(_ chapterId: String) throws {
        try callWithDataInput(Data(chapterId.utf8)) { inPtr, inLen in
            bb_unload_chapter(handle, inPtr, inLen)
        }
    }

    public func serializeState() throws -> String {
        try String(decoding: callWithOutput { outPtr, outCap in
            bb_serialize(handle, outPtr, outCap)
        }, as: UTF8.self)
    }

    public func restoreState(_ json: String) throws -> String {
        try String(decoding: callWithUtf8Input(json) { inPtr, inLen, outPtr, outCap in
            bb_restore(handle, inPtr, inLen, outPtr, outCap)
        }, as: UTF8.self)
    }

    private struct Copy {
        let ptr: Int
        let len: UInt32
    }

    private static func copyIntoRust(_ data: Data) throws -> Copy {
        let len = UInt32(data.count)
        guard len > 0 else {
            return Copy(ptr: 0, len: 0)
        }
        let ptr = bb_alloc(len)
        guard ptr != 0 else {
            throw BlackboxEngineError.allocFailed
        }
        _ = data.withUnsafeBytes { raw in
            memcpy(UnsafeMutableRawPointer(bitPattern: ptr)!, raw.baseAddress!, raw.count)
        }
        return Copy(ptr: ptr, len: len)
    }

    private static func freeCopy(_ copy: Copy) {
        if copy.ptr != 0 {
            bb_free(copy.ptr, copy.len)
        }
    }

    private static func readLastError() -> String {
        var cap: UInt32 = 512
        while cap <= 65_536 {
            let allocatedCap = cap
            let ptr = bb_alloc(allocatedCap)
            guard ptr != 0 else {
                return "bb_alloc failed while reading last error"
            }
            defer { bb_free(ptr, allocatedCap) }
            let written = bb_last_error(ptr, allocatedCap)
            if written > 0 {
                let bytes = Data(bytes: UnsafeRawPointer(bitPattern: ptr)!, count: Int(written))
                return String(decoding: bytes, as: UTF8.self)
            }
            if written < 0 {
                cap = UInt32(-written)
                continue
            }
            return "unknown blackbox error"
        }
        return "blackbox error message exceeded buffer limit"
    }

    private func callWithOutput(
        _ executor: (_ outPtr: Int, _ outCap: UInt32) -> Int32
    ) throws -> Data {
        var cap: UInt32 = 4096
        while cap <= 4 * 1024 * 1024 {
            let allocatedCap = cap
            let outPtr = bb_alloc(allocatedCap)
            guard outPtr != 0 else {
                throw BlackboxEngineError.allocFailed
            }
            defer { bb_free(outPtr, allocatedCap) }
            let written = executor(outPtr, allocatedCap)
            if written > 0 {
                return Data(bytes: UnsafeRawPointer(bitPattern: outPtr)!, count: Int(written))
            }
            if written < 0 {
                cap = UInt32(-written)
                continue
            }
            throw BlackboxEngineError.callFailed(Self.readLastError())
        }
        throw BlackboxEngineError.outputTooLarge
    }

    private func callWithUtf8Input(
        _ input: String,
        _ executor: (_ inPtr: Int, _ inLen: UInt32, _ outPtr: Int, _ outCap: UInt32) -> Int32
    ) throws -> Data {
        let copied = try Self.copyIntoRust(Data(input.utf8))
        defer { Self.freeCopy(copied) }
        return try callWithOutput { outPtr, outCap in
            executor(copied.ptr, copied.len, outPtr, outCap)
        }
    }

    private func callWithDataInput(
        _ input: Data,
        _ executor: (_ inPtr: Int, _ inLen: UInt32) -> Int32
    ) throws {
        let copied = try Self.copyIntoRust(input)
        defer { Self.freeCopy(copied) }
        guard executor(copied.ptr, copied.len) != 0 else {
            throw BlackboxEngineError.callFailed(Self.readLastError())
        }
    }
}

File: apps/android/blackbox/src/main/kotlin/com/blackbox/engine/BlackboxEngine.kt
package com.blackbox.engine

class BlackboxEngine private constructor(private val handle: Int) : AutoCloseable {
    init {
        require(handle != 0) { "invalid engine handle" }
    }

    override fun close() {
        Native.bbEngineFree(handle)
    }

    fun getCurrentView(): String = Native.callWithOutput { outPtr, outCap ->
        Native.bbGetView(handle, outPtr, outCap)
    }

    fun submitCommand(json: String, viewRevision: Int): String = Native.callWithUtf8Input(json) { inPtr, inLen, outPtr, outCap ->
        Native.bbSubmit(handle, inPtr, inLen, viewRevision, outPtr, outCap)
    }

    fun loadCatalog(bytes: ByteArray) = Native.callWithBytesInput(bytes) { ptr, len ->
        Native.bbLoadCatalog(handle, ptr, len)
    }

    fun loadLibrary(bytes: ByteArray) = Native.callWithBytesInput(bytes) { ptr, len ->
        Native.bbLoadLibrary(handle, ptr, len)
    }

    fun loadChapter(bytes: ByteArray) = Native.callWithBytesInput(bytes) { ptr, len ->
        Native.bbLoadChapter(handle, ptr, len)
    }

    fun unloadChapter(chapterId: String) =
        Native.callWithBytesInput(chapterId.toByteArray(Charsets.UTF_8)) { ptr, len ->
            Native.bbUnloadChapter(handle, ptr, len)
        }

    fun serializeState(): String = Native.callWithOutput { outPtr, outCap ->
        Native.bbSerialize(handle, outPtr, outCap)
    }

    fun restoreState(json: String): String = Native.callWithUtf8Input(json) { inPtr, inLen, outPtr, outCap ->
        Native.bbRestore(handle, inPtr, inLen, outPtr, outCap)
    }

    companion object {
        fun initialize() {
            Native.bbInit()
        }

        fun create(
            scenario: ByteArray,
            items: ByteArray,
            characters: ByteArray,
            assets: ByteArray,
            chapters: List<ByteArray>,
            library: ByteArray? = null,
            randomSeedOverride: ULong? = null,
        ): BlackboxEngine {
            val scenarioCopy = Native.copyIntoRust(scenario)
            val itemsCopy = Native.copyIntoRust(items)
            val charactersCopy = Native.copyIntoRust(characters)
            val assetsCopy = Native.copyIntoRust(assets)
            val libraryCopy = Native.copyIntoRust(library ?: byteArrayOf())
            val chapterCopies = chapters.map { Native.copyIntoRust(it) }

            val tablePtr: Long
            val tableLen: Int
            if (chapterCopies.isEmpty()) {
                tablePtr = 0L
                tableLen = 0
            } else {
                val stride = 16 // matches #[repr(C)] ByteSlice on 64-bit (usize + u32 + padding)
                val table = ByteArray(chapterCopies.size * stride)
                var offset = 0
                for (copy in chapterCopies) {
                    writePtr(table, offset, copy.ptr)
                    writeU32(table, offset + 8, copy.len)
                    offset += stride
                }
                tableLen = table.size
                tablePtr = Native.copyIntoRust(table).ptr
            }

            try {
                val handle = Native.bbEngineNew(
                    scenarioCopy.ptr,
                    scenarioCopy.len,
                    itemsCopy.ptr,
                    itemsCopy.len,
                    charactersCopy.ptr,
                    charactersCopy.len,
                    assetsCopy.ptr,
                    assetsCopy.len,
                    tablePtr,
                    chapters.size,
                    libraryCopy.ptr,
                    libraryCopy.len,
                    randomSeedOverride != null,
                    randomSeedOverride?.toLong() ?: 0L,
                )
                if (handle == 0) {
                    error(Native.readLastError())
                }
                return BlackboxEngine(handle)
            } finally {
                Native.freeCopy(scenarioCopy)
                Native.freeCopy(itemsCopy)
                Native.freeCopy(charactersCopy)
                Native.freeCopy(assetsCopy)
                Native.freeCopy(libraryCopy)
                for (copy in chapterCopies) {
                    Native.freeCopy(copy)
                }
                if (tablePtr != 0L) {
                    Native.bbFree(tablePtr, tableLen)
                }
            }
        }

        private fun writePtr(buffer: ByteArray, offset: Int, value: Long) {
            for (shift in 0 until 8) {
                buffer[offset + shift] = ((value ushr (8 * shift)) and 0xFF).toByte()
            }
        }

        private fun writeU32(buffer: ByteArray, offset: Int, value: Int) {
            for (shift in 0 until 4) {
                buffer[offset + shift] = ((value ushr (8 * shift)) and 0xFF).toByte()
            }
        }
    }
}

internal object Native {
    init {
        System.loadLibrary("blackbox_ffi")
        System.loadLibrary("blackbox_jni")
    }

    data class Copy(val ptr: Long, val len: Int)

    external fun bbInit()
    external fun bbAlloc(len: Int): Long
    external fun bbFree(ptr: Long, len: Int)
    external fun bbEngineNew(
        scenarioPtr: Long,
        scenarioLen: Int,
        itemsPtr: Long,
        itemsLen: Int,
        charactersPtr: Long,
        charactersLen: Int,
        assetsPtr: Long,
        assetsLen: Int,
        chaptersPtr: Long,
        chapterCount: Int,
        libraryPtr: Long,
        libraryLen: Int,
        hasRandomSeedOverride: Boolean,
        randomSeedOverride: Long,
    ): Int

    external fun bbEngineFree(handle: Int)
    external fun bbGetView(handle: Int, outPtr: Long, outCap: Int): Int
    external fun bbSubmit(
        handle: Int,
        inPtr: Long,
        inLen: Int,
        viewRevision: Int,
        outPtr: Long,
        outCap: Int,
    ): Int
    external fun bbSerialize(handle: Int, outPtr: Long, outCap: Int): Int
    external fun bbRestore(handle: Int, inPtr: Long, inLen: Int, outPtr: Long, outCap: Int): Int
    external fun bbLastError(outPtr: Long, outCap: Int): Int
    external fun bbLoadCatalog(handle: Int, inPtr: Long, inLen: Int): Int
    external fun bbLoadLibrary(handle: Int, inPtr: Long, inLen: Int): Int
    external fun bbLoadChapter(handle: Int, inPtr: Long, inLen: Int): Int
    external fun bbUnloadChapter(handle: Int, inPtr: Long, inLen: Int): Int

    fun copyIntoRust(bytes: ByteArray): Copy {
        if (bytes.isEmpty()) {
            return Copy(0L, 0)
        }
        val ptr = bbAlloc(bytes.size)
        require(ptr != 0L) { "bb_alloc failed" }
        copyBytes(ptr, bytes)
        return Copy(ptr, bytes.size)
    }

    fun freeCopy(copy: Copy) {
        if (copy.ptr != 0L) {
            bbFree(copy.ptr, copy.len)
        }
    }

    private external fun copyBytes(ptr: Long, bytes: ByteArray)

    fun readLastError(): String = callWithOutput { outPtr, outCap ->
        bbLastError(outPtr, outCap)
    }

    fun callWithOutput(executor: (outPtr: Long, outCap: Int) -> Int): String {
        var cap = 4096
        while (cap <= 4 * 1024 * 1024) {
            val allocatedCap = cap
            val outPtr = bbAlloc(allocatedCap)
            require(outPtr != 0L) { "bb_alloc failed" }
            try {
                val written = executor(outPtr, allocatedCap)
                when {
                    written > 0 -> return decode(outPtr, written)
                    written < 0 -> cap = -written
                    else -> error(readLastError())
                }
            } finally {
                bbFree(outPtr, allocatedCap)
            }
        }
        error("blackbox output exceeded maximum buffer size")
    }

    fun callWithUtf8Input(
        input: String,
        executor: (inPtr: Long, inLen: Int, outPtr: Long, outCap: Int) -> Int,
    ): String {
        val bytes = input.toByteArray(Charsets.UTF_8)
        val copy = copyIntoRust(bytes)
        try {
            return callWithOutput { outPtr, outCap ->
                executor(copy.ptr, copy.len, outPtr, outCap)
            }
        } finally {
            freeCopy(copy)
        }
    }

    fun callWithBytesInput(bytes: ByteArray, executor: (inPtr: Long, inLen: Int) -> Int) {
        val copy = copyIntoRust(bytes)
        try {
            if (executor(copy.ptr, copy.len) == 0) error(readLastError())
        } finally {
            freeCopy(copy)
        }
    }

    private external fun decode(ptr: Long, len: Int): String
}
